deepali/core/image.py

```diff
@@ -1163,6 +1164,81 @@ def grid_sample_mask(
     )
 
 
+def rand_sample(
+    data: Union[Tensor, Sequence[Tensor]],
+    num_samples: int,
+    mask: Optional[Tensor] = None,
+    replacement: bool = False,
+    generator: Optional[Generator] = None,
+) -> Union[Tensor, Sequence[Tensor]]:
+    r"""Random sampling of voxels within an image.
+
+    Args:
+        data: One or more image batch tensors with shape ``(N, C, ... X)`` to sample
+            from at the same random image grid positions, i.e., voxel indices. Note that
+            all input tensors must have the same number and size of spatial dimensions.
+        num_samples: Number of spatial samples to draw.
+        mask: Optional mask to use for spatially weighted sampling.
+        replacement: Whether to sample with or without replacement.
+        generator: Random number generator to use.
+
+    Returns:
+        Tensor of shape ``(N, C, num_samples)`` with input ``data`` values at randomly
+        sampled spatial grid points. When ``data`` is a sequence of tensors, a list
+        of tensors with order matching the input data is returned.
+
+    """
+    input: Sequence[Tensor]
+    if isinstance(data, Tensor):
+        input = [data]
+    elif isinstance(data, Sequence) and all(isinstance(x, Tensor) for x in data):
+        input = data
+    else:
+        raise TypeError("rand_sample() 'data' must be Tensor or Sequence[Tensor]")
+    if not input:
+        return []
+    if any(x.ndim < 3 for x in input):
+        raise ValueError("rand_sample() 'data' must be one or more tensors of shape (N, C, ..., X)")
+    shape = input[0].shape
+    device = input[0].device
+    if any(x.shape[2:] != shape[2:] for x in input):
+        raise ValueError("rand_sample() 'data' tensors must have identical spatial shape")
+    numel = shape[2:].numel()
+    if not replacement and num_samples > numel:
+        raise ValueError("rand_sample() 'num_samples' is greater than number of spatial points")
+    input = [x.flatten(2) for x in input]
+    if mask is None:
```

Comment:
Right now the indices for different tensors in a batch would be the same when sampling without a mask and without replacement, while different for each tensor in other settings. Would it make sense to use `multinomial()` with a mask of equal weights for the no-mask setting? 
This shouldn't make much difference function-wise, but it would make the behavior more consistent and the code cleaner.

Reply:
Simplifying the code by always using `multinomial` with a default mask of ones makes sense. It would probably be best to time this. Note that the current implementation does indeed sample different indices for different samples in a batch for all code paths. Notice the `for row in index` loop of this modified code which produces a different permutation for each sample in the batch.

Reply:
I've done some simple timing measurements, and seems because of the loop `randperm` is considerably slower than just using `multinomial` with a mask of ones:

```
t_elapsed_1 = 0
for i in range(5):
    start.record()
    indices = U.rand_sample(data, num_samples, mask=None, replacement=False, generator=generator)
    end.record()
    torch.cuda.synchronize()
    if i > 0:
        t_elapsed_1 += start.elapsed_time(end)
t_elapsed_1 /= 4

t_elapsed_2 = 0
for i in range(5):
    start.record()
    mask = torch.ones((1, 1) + data.shape[2:], device=device)
    indices = U.rand_sample(data, num_samples, mask=mask, replacement=False, generator=generator)
    end.record()
    torch.cuda.synchronize()
    if i > 0:
        t_elapsed_2 += start.elapsed_time(end)
t_elapsed_2 /= 4
```

Gives
- t_elapsed_1 = 1.44 ms
- t_elapsed_2 = 0.35 ms


In case of `replacement=True`, calling `randint` is the simplest and fastest, though. So probably best to keep that `if` branch.

Reply:
For an input of larger size, hitting the limit of `torch.multinomial`, things look differently:

Input size `(5, 1, 256, 512, 512)`:
- t_elapsed_time_1 = 110 ms
- t_elapsed_time_2 = 6277 ms

The inverse transform sampling is considerably more expensive than a few `randperm` calls.

Reply:
Could also consider using https://github.com/LeviViana/torch_sampling which may be more performant.

Reply:
Yes sorry I misread the randperm-copy loop. Indeed it's generating different index for different samples. I agree with the conclusion with what's shown from timing.